Indent level (parameters: first) for async arrow functions

Probably the same as #10531 which was automatically closed due to inactivity 2 years ago.

**Tell us about your environment**

ESLint Version: **7.4.0**
Node Version: **14.5.0**
yarn Version: **1.22.4**
What parser (default, Babel-ESLint, etc.) are you using? **espree & @typescript-eslint/parser**

**Configuration**

```javascript
const indentOptions = {
  "FunctionDeclaration": { "parameters": "first" },
  "FunctionExpression": { "parameters": "first" },
};
const indent = ["error", 2, indentOptions];

module.exports = {
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: "module"
  },
  overrides: [
    {
      files: ["*.js"],
      parser: "espree",
      rules: {
        indent: indent
      }
    }
  ]
};
```

**What did you do?**

```javascript
const foo = async (arg1,
                   arg2) =>
{
  return arg1 + arg2;
}
```

```node_modules/.bin/eslint foo.js```

**What did you expect to happen?**

Run without errors

**What actually happened?**

```bash
/private/tmp/foo/foo.js
  2:1  error  Expected indentation of 2 spaces but found 19  indent

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

**Note:** if I remove the async then the error goes away (the "parameters": "first" is honoured).

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -1094,6 +1094,17 @@ module.exports = {
                     parameterParens.add(closingParen);
                     addElementListIndent(node.params, openingParen, closingParen, options.FunctionExpression.parameters);
                 }
+
+                if (firstToken.type === "Identifier" && firstToken.value === "async") {
+                    const openingParen = sourceCode.getTokenAfter(firstToken);
+                    const closingParen = sourceCode.getTokenBefore(node.body, astUtils.isClosingParenToken);
+
+                    parameterParens.add(openingParen);
+                    parameterParens.add(closingParen);
+
+                    addElementListIndent(node.params, openingParen, closingParen, options.FunctionExpression.parameters);
+                }
+
                 addBlocklessNodeIndent(node.body);
             },
 

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -1084,27 +1084,15 @@ module.exports = {
             },
 
             ArrowFunctionExpression(node) {
-                const firstToken = sourceCode.getFirstToken(node);
+                const maybeOpeningParen = sourceCode.getFirstToken(node, node.async ? 1 : 0);
 
-                if (astUtils.isOpeningParenToken(firstToken)) {
-                    const openingParen = firstToken;
+                if (astUtils.isOpeningParenToken(maybeOpeningParen)) {
+                    const openingParen = maybeOpeningParen;
                     const closingParen = sourceCode.getTokenBefore(node.body, astUtils.isClosingParenToken);
 
                     parameterParens.add(openingParen);
                     parameterParens.add(closingParen);
                     addElementListIndent(node.params, openingParen, closingParen, options.FunctionExpression.parameters);
-                }
-
-                if (firstToken.type === "Identifier" && firstToken.value === "async") {
-                    const nextToken = sourceCode.getTokenAfter(firstToken);
-
-                    if (astUtils.isOpeningParenToken(nextToken)) {
-                        const closingParen = sourceCode.getTokenBefore(node.body, astUtils.isClosingParenToken);
-
-                        parameterParens.add(nextToken);
-                        parameterParens.add(closingParen);
-                        addElementListIndent(node.params, nextToken, closingParen, options.FunctionExpression.parameters);
-                    }
                 }
 
                 addBlocklessNodeIndent(node.body);

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -1097,12 +1097,14 @@ module.exports = {
 
                 if (firstToken.type === "Identifier" && firstToken.value === "async") {
                     const openingParen = sourceCode.getTokenAfter(firstToken);
-                    const closingParen = sourceCode.getTokenBefore(node.body, astUtils.isClosingParenToken);
 
-                    parameterParens.add(openingParen);
-                    parameterParens.add(closingParen);
+                    if (astUtils.isOpeningParenToken(openingParen)) {
+                        const closingParen = sourceCode.getTokenBefore(node.body, astUtils.isClosingParenToken);
 
-                    addElementListIndent(node.params, openingParen, closingParen, options.FunctionExpression.parameters);
+                        parameterParens.add(openingParen);
+                        parameterParens.add(closingParen);
+                        addElementListIndent(node.params, openingParen, closingParen, options.FunctionExpression.parameters);
+                    }
                 }
 
                 addBlocklessNodeIndent(node.body);

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -1084,7 +1084,7 @@ module.exports = {
             },
 
             ArrowFunctionExpression(node) {
-                const maybeOpeningParen = sourceCode.getFirstToken(node, node.async ? 1 : 0);
+                const maybeOpeningParen = sourceCode.getFirstToken(node, { skip: node.async ? 1 : 0 });
 
                 if (astUtils.isOpeningParenToken(maybeOpeningParen)) {
                     const openingParen = maybeOpeningParen;

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -1096,14 +1096,14 @@ module.exports = {
                 }
 
                 if (firstToken.type === "Identifier" && firstToken.value === "async") {
-                    const openingParen = sourceCode.getTokenAfter(firstToken);
+                    const nextToken = sourceCode.getTokenAfter(firstToken);
 
-                    if (astUtils.isOpeningParenToken(openingParen)) {
+                    if (astUtils.isOpeningParenToken(nextToken)) {
                         const closingParen = sourceCode.getTokenBefore(node.body, astUtils.isClosingParenToken);
 
-                        parameterParens.add(openingParen);
+                        parameterParens.add(nextToken);
                         parameterParens.add(closingParen);
-                        addElementListIndent(node.params, openingParen, closingParen, options.FunctionExpression.parameters);
+                        addElementListIndent(node.params, nextToken, closingParen, options.FunctionExpression.parameters);
                     }
                 }
 

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -11600,8 +11600,8 @@ ruleTester.run("indent", rule, {
                  b => {}
             `,
             output: unIndent`
-            const a = async
-            b => {}
+                const a = async
+                b => {}
             `,
             options: [2],
             errors: expectedErrors([

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -5732,7 +5732,7 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
@@ -5742,14 +5742,14 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
                 const a = async
                 b => {}
-                `,
+            `,
             options: [2]
         },
         {
@@ -5760,7 +5760,7 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-                    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
@@ -5770,7 +5770,7 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-                    `,
+            `,
             options: [2]
         },
         {
@@ -5780,7 +5780,7 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-                    `,
+            `,
             options: [2]
         },
         {
@@ -5790,7 +5790,7 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-                    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: 4 }, FunctionExpression: { parameters: 4 } }]
         },
         {
@@ -5800,28 +5800,28 @@ ruleTester.run("indent", rule, {
                 {
                   return arg1 + arg2;
                 }
-                    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: 4 }, FunctionExpression: { parameters: 4 } }]
         },
         {
             code: unIndent`
                 async function fn(ar1,
                                   ar2){}
-                    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
                 async function /* some comments */ fn(ar1,
                                                       ar2){}
-                    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
                 async  /* some comments */  function fn(ar1,
                                                         ar2){}
-                    `,
+            `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         }
     ],
@@ -11575,18 +11575,18 @@ ruleTester.run("indent", rule, {
         },
         {
             code: unIndent`
-const foo = async (arg1,
-                    arg2) =>
-{
-  return arg1 + arg2;
-}
+                const foo = async (arg1,
+                                    arg2) =>
+                {
+                  return arg1 + arg2;
+                }
             `,
             output: unIndent`
-const foo = async (arg1,
-                   arg2) =>
-{
-  return arg1 + arg2;
-}
+                const foo = async (arg1,
+                                   arg2) =>
+                {
+                  return arg1 + arg2;
+                }
             `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }],
             parserOptions: { ecmaVersion: 2020 },
@@ -11596,13 +11596,13 @@ const foo = async (arg1,
         },
         {
             code: unIndent`
-const a = async
- b => {}
-`,
+                const a = async
+                 b => {}
+            `,
             output: unIndent`
-const a = async
-b => {}
-`,
+            const a = async
+            b => {}
+            `,
             options: [2],
             errors: expectedErrors([
                 [2, 0, 1, "Identifier"]

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -5822,6 +5822,13 @@ async function /* some comments */ fn(ar1,
                                       ar2){}
     `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
+        },
+        {
+            code: unIndent`
+async  /* some comments */  function fn(ar1,
+                                        ar2){}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         }
     ],
 

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -5724,6 +5724,64 @@ ruleTester.run("indent", rule, {
             `,
             options: [4, { MemberExpression: 2 }],
             parserOptions: { ecmaVersion: 2015 }
+        },
+        {
+            code: unIndent`
+const foo = async (arg1,
+                   arg2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
+        },
+        {
+            code: unIndent`
+const foo = (arg1,
+             arg2) => async (arr1,
+                             arr2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
+        },
+        {
+            code: unIndent`
+const foo = async (arg1,
+  arg2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2]
+        },
+        {
+            code: unIndent`
+const foo = async (arg1,
+        arg2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: 4 }, FunctionExpression: { parameters: 4 } }]
+        },
+        {
+            code: unIndent`
+const foo = (arg1,
+        arg2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: 4 }, FunctionExpression: { parameters: 4 } }]
+        },
+        {
+            code: unIndent`
+async function fn(ar1,
+                  ar2){}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         }
     ],
 
@@ -11472,6 +11530,27 @@ ruleTester.run("indent", rule, {
             errors: expectedErrors([
                 [6, 4, 0, "Punctuator"],
                 [7, 4, 0, "Punctuator"]
+            ])
+        },
+        {
+            code: unIndent`
+const foo = async (arg1,
+                    arg2) =>
+{
+  return arg1 + arg2;
+}
+            `,
+            output: unIndent`
+const foo = async (arg1,
+                   arg2) =>
+{
+  return arg1 + arg2;
+}
+            `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }],
+            parserOptions: { ecmaVersion: 2020 },
+            errors: expectedErrors([
+                [2, 19, 20, "Identifier"]
             ])
         }
     ]

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -5737,7 +5737,17 @@ const foo = async (arg1,
         },
         {
             code: unIndent`
-const a = async b => {}
+const foo = async /* some comments */(arg1,
+                                      arg2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
+        },
+        {
+            code: unIndent`
+const a = async /*comments */b => {}
     `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
@@ -5771,6 +5781,16 @@ const foo = async (arg1,
         },
         {
             code: unIndent`
+const foo = async /*comments*/(arg1,
+  arg2) =>
+{
+  return arg1 + arg2;
+}
+    `,
+            options: [2]
+        },
+        {
+            code: unIndent`
 const foo = async (arg1,
         arg2) =>
 {
@@ -5793,6 +5813,13 @@ const foo = (arg1,
             code: unIndent`
 async function fn(ar1,
                   ar2){}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
+        },
+        {
+            code: unIndent`
+async function /* some comments */ fn(ar1,
+                                      ar2){}
     `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         }

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -5737,6 +5737,19 @@ const foo = async (arg1,
         },
         {
             code: unIndent`
+const a = async b => {}
+    `,
+            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
+        },
+        {
+            code: unIndent`
+const a = async
+b => {}
+`,
+            options: [2]
+        },
+        {
+            code: unIndent`
 const foo = (arg1,
              arg2) => async (arr1,
                              arr2) =>
@@ -11551,6 +11564,20 @@ const foo = async (arg1,
             parserOptions: { ecmaVersion: 2020 },
             errors: expectedErrors([
                 [2, 19, 20, "Identifier"]
+            ])
+        },
+        {
+            code: unIndent`
+const a = async
+ b => {}
+`,
+            output: unIndent`
+const a = async
+b => {}
+`,
+            options: [2],
+            errors: expectedErrors([
+                [2, 0, 1, "Identifier"]
             ])
         }
     ]

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -5727,107 +5727,101 @@ ruleTester.run("indent", rule, {
         },
         {
             code: unIndent`
-const foo = async (arg1,
-                   arg2) =>
-{
-  return arg1 + arg2;
-}
+                const foo = async (arg1,
+                                   arg2) =>
+                {
+                  return arg1 + arg2;
+                }
     `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
-const foo = async /* some comments */(arg1,
-                                      arg2) =>
-{
-  return arg1 + arg2;
-}
+                const foo = async /* some comments */(arg1,
+                                                      arg2) =>
+                {
+                  return arg1 + arg2;
+                }
     `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
-const a = async /*comments */b => {}
-    `,
-            options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
-        },
-        {
-            code: unIndent`
-const a = async
-b => {}
-`,
+                const a = async
+                b => {}
+                `,
             options: [2]
         },
         {
             code: unIndent`
-const foo = (arg1,
-             arg2) => async (arr1,
-                             arr2) =>
-{
-  return arg1 + arg2;
-}
-    `,
+                const foo = (arg1,
+                             arg2) => async (arr1,
+                                             arr2) =>
+                {
+                  return arg1 + arg2;
+                }
+                    `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
-const foo = async (arg1,
-  arg2) =>
-{
-  return arg1 + arg2;
-}
-    `,
+                const foo = async (arg1,
+                  arg2) =>
+                {
+                  return arg1 + arg2;
+                }
+                    `,
             options: [2]
         },
         {
             code: unIndent`
-const foo = async /*comments*/(arg1,
-  arg2) =>
-{
-  return arg1 + arg2;
-}
-    `,
+                const foo = async /*comments*/(arg1,
+                  arg2) =>
+                {
+                  return arg1 + arg2;
+                }
+                    `,
             options: [2]
         },
         {
             code: unIndent`
-const foo = async (arg1,
-        arg2) =>
-{
-  return arg1 + arg2;
-}
-    `,
+                const foo = async (arg1,
+                        arg2) =>
+                {
+                  return arg1 + arg2;
+                }
+                    `,
             options: [2, { FunctionDeclaration: { parameters: 4 }, FunctionExpression: { parameters: 4 } }]
         },
         {
             code: unIndent`
-const foo = (arg1,
-        arg2) =>
-{
-  return arg1 + arg2;
-}
-    `,
+                const foo = (arg1,
+                        arg2) =>
+                {
+                  return arg1 + arg2;
+                }
+                    `,
             options: [2, { FunctionDeclaration: { parameters: 4 }, FunctionExpression: { parameters: 4 } }]
         },
         {
             code: unIndent`
-async function fn(ar1,
-                  ar2){}
-    `,
+                async function fn(ar1,
+                                  ar2){}
+                    `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
-async function /* some comments */ fn(ar1,
-                                      ar2){}
-    `,
+                async function /* some comments */ fn(ar1,
+                                                      ar2){}
+                    `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         },
         {
             code: unIndent`
-async  /* some comments */  function fn(ar1,
-                                        ar2){}
-    `,
+                async  /* some comments */  function fn(ar1,
+                                                        ar2){}
+                    `,
             options: [2, { FunctionDeclaration: { parameters: "first" }, FunctionExpression: { parameters: "first" } }]
         }
     ],

